llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp

```diff
@@ -112,7 +112,7 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
 
           // Reset uses count when a register is no longer live.
           for (const MCRegUnit &Unit : RegUnits)
-            RegisterUseCount.erase(Unit);
+            RegisterUseCount[Unit] = 0;
```

Comment:
It's really hard to understand the effect of this change because of other problems in this loop-over-the-operands.

1. The `modifiesRegister` check on line 104 is odd because it checks if _any_ operand of the current instruction modifies Reg, not just the current operand. Can it just check `Operand.isDef()` instead?

2. Really the loop should be split into two processing defs first and then uses. I think you had a downstream patch to do that.

Could you post a patch to fix either (1) or (2) first, and then we can revisit this one?